---
title: "Age Stratification - BRAIN"
author: "Sonia Garcia-Ruiz"
date: "27/03/2022"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
    #number_sections: true
    theme: flatly
    highlight: tango
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(rutils)

project_id <- "BRAIN"

source(paste0("/home/sruiz/PROJECTS/splicing-project/pipeline3_age_stratification.R"))

ref <- rtracklayer::import(con = "/data/references/ensembl/gtf/v105/Homo_sapiens.GRCh38.105.chr.gtf")

age_samples_clusters_tidy <- age_stratification_init_data(project_id)
age_supergroups <- age_samples_clusters_tidy$age_group %>% unique()


folder_root <- paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/", project_id,
                      "/results/pipeline3/missplicing-ratio/age/")

## Load the IDBs
df_age_groups_intron <- map_df(age_supergroups, function(age_group) {
  
  # age_group <- age_supergroups[1]
  readRDS(file = paste0(folder_root, "/", age_group, "/", age_group, "_db_introns.rds")) %>%
    #mutate(ref_junID = ref_junID %>% as.integer()) %>%
    mutate(sample_type = age_group) %>%
    return()
  
})
df_age_groups_novel <- map_df(age_supergroups, function(age_group) {
  
  #print(paste0(Sys.time(), " - loading IDB for '", age_group, "' samples ..."))
  readRDS(file = paste0(folder_root, "/", age_group, "/", age_group, "_db_novel.rds")) %>%
    #mutate(ref_junID = ref_junID %>% as.integer()) %>%
    #mutate(novel_junID = novel_junID %>% as.integer()) %>%
    mutate(sample_type = age_group) %>%
    return()
  
}) 

## Get the junctions that are common across age supergroups

common_introns_misspliced <- df_age_groups_novel %>%
  group_by(sample_type) %>%
  distinct(ref_junID, .keep_all = T) %>%
  ungroup() %>%
  dplyr::count(ref_junID) %>%
  filter(n == age_supergroups %>% length()) %>%
  pull(ref_junID)

common_introns <- df_age_groups_intron %>%
  group_by(sample_type) %>%
  distinct(ref_junID, .keep_all = T) %>%
  ungroup() %>%
  dplyr::count(ref_junID) %>%
  filter(n == age_supergroups %>% length()) %>%
  pull(ref_junID)

# setdiff(common_introns_misspliced,common_introns)
# setdiff(common_introns,common_introns_misspliced)
# common_introns_misspliced %>% length()
# common_introns %>% length()
# intersect(common_introns_misspliced,common_introns) %>% length()

## Filter by the common junctions
df_age_groups_intron_tidy <- merge(x = df_age_groups_intron %>% data.table::as.data.table(),
                                   y = data.table::data.table(ref_junID = common_introns_misspliced),
                                   by = "ref_junID",
                                   all.y = T)



df_age_groups_novel_tidy <- merge(x = df_age_groups_novel %>% data.table::as.data.table(),
                                  y = data.table::data.table(ref_junID = common_introns_misspliced),
                                  by = "ref_junID",
                                  all.y = T)



bg_genes <- df_age_groups_novel_tidy %>% distinct(gene_name) %>% pull() %>% unlist %>% unique()
# bg_genes2 <- df_age_groups_intron_tidy %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
# 
# 
# bg_genes %>% length()
# bg_genes2 %>% length()
# identical(bg_genes,bg_genes2)

```

# Data used

This document contains the age analysis performed using RNA-seq data from the GTEx v8 project and processed by [recount3](https://jhubiostatistics.shinyapps.io/recount3-study-explorer/).

Samples were grouped together within the following age clusters:

* 20-39: contains samples from the age groups "20-29" and "30-39"
* 40-59: contains samples from the age groups "40-49" and "50-59"
* 60-79: contains samples from the age groups "60-69" and "70-79".


# Stats table {.tabset .tabset-fade .tabset-pills}

Samples have been displayed after being grouped by different columns such as 'age', 'sex', 'RIN' and 'body region'.

## Across tissues

```{r mean_values}

age_samples_clusters_tidy %>%
  select(-individual, -region, -age, -sample_recount_id, -region) %>%
  group_by(age_group) %>%
  mutate(median_rin = rin %>% median())%>%
  mutate(mean_read_depth = round(mapped_read_count %>% mean())) %>%
  select(-rin,-mapped_read_count) %>%
  mutate(N_samples = n()) %>%
  ungroup() %>%
  #group_by(region, age_group) %>%
  #mutate(N_samples_tissue = n()) %>%
  #ungroup() %>%
  group_by(age_group, sex) %>%
  mutate(sex_ratio = round(n()/N_samples, digits = 2)) %>%
  distinct(age_group, .keep_all = T) %>%
  spread(key = sex, value = sex_ratio) %>%
  dplyr::rename(ratio_females = female,
                ratio_males = male) %>%
  relocate(N_samples, .after = age_group) %>%
  DT::datatable(rownames = F)
  
```


Column description:

* **N_samples:** total number of samples clustered within the age super-group.
* **median_rin:** this is the median RIN number of all samples clustered within the age group. 
* **mean_read_depth**: this is the mean 'mapped_read_depth' of all samples clustered within the age group. 
* **ratio_females:** this is the ratio of female samples available within the age group.
* **ratio_males:** this is the ratio of male samples available within the age group.


## Per tissue

```{r brain_region}

age_samples_clusters_tidy %>% as_tibble() %>%
  select(-individual, -age, -sex,-sample_recount_id) %>%
  group_by(age_group) %>%
  mutate(median_rin = rin %>% median()) %>%
  mutate(mean_read_depth = round(mapped_read_count %>% mean())) %>%
  select(-rin,-mapped_read_count) %>%
  mutate(N_samples = n()) %>%
  ungroup() %>%
    
  group_by(region, age_group) %>%
  mutate(N_samples_tissue = n()) %>%
  ungroup() %>%

  # group_by(age_group, sex) %>%
  # mutate(sex_ratio = round(x = n()/N_samples,
  #                          digits = 2)) %>%
  # ungroup() %>%
  
  group_by(region, age_group) %>%
  distinct(age_group, .keep_all = T) %>%
  #spread(key = sex, value = sex_ratio) %>%
  dplyr::arrange(region) %>%
    #ungroup() %>%
    #group_by(age_group) %>%
    #mutate(mean_samples_across_tissues = N_samples_tissue %>% mean) %>%
    #mutate(median_samples_across_tissues = N_samples_tissue %>% median) %>%
  spread(key = region, value = N_samples_tissue) %>% 
    #select(-region, -N_samples_tissue) %>%
  distinct(age_group, .keep_all = T) %>% 
  #spread(key = sex, value = sex_ratio)  %>%
  as.data.frame()  %>%
  # dplyr::rename(ratio_females = female,
  #               ratio_males = male) %>%
  relocate(N_samples, .after = age_group) %>%
  DT::datatable(rownames = F)
  
```


Column description:

* **N_samples:** total number of samples clustered within the age group.
* **median_rin:** this is the median RIN number of all samples clustered within the age group. 
* **mean_read_depth**: this is the mean 'mapped_read_depth' of all samples clustered within the age group. 
* **tissue**: This is the exact number of samples available per body region clustered within each age group.


# Plots {.tabset .tabset-fade .tabset-pills}

## Distances

<!-- In this graph it has only been included **116,847** reference introns that were common across all three age supergroups. -->
<!-- The plot below shows the distances from all novel junction located within a window of -60/+60 bp to their reference introns.  -->

In this graph it has only been included common reference introns across all three age supergroups.
The plot below shows the distances from all novel junction located within a window of -30/+30 bp from their reference introns.

Interestingly, the distances peak at the donor exon location is not there anymore if we compare these results with the ones obtained using GTEx v6 data. This is the same type of distances result that we obtained using data from the ENCODE RBP knockdown projects. 


```{r distances_plot, warning=FALSE}

age_stratification_plot_distances(df = df_age_groups_novel_tidy, 
                                  age_levels = c("60-79","40-59","20-39" ), 
                                  distance_limit = 60) %>% print()

```

ENCODE Pipeline overview:

* Illumina TruSeq protocol to create a polyA selected strand specific library.
* RNA populations ~200bp 
* Mapping of the reads to the genome using STAR.
* Gene annotation files using GENCODE

Recount3/GTEx v8 pipeline overview:

* The Illumina TruSeq protocol was used to create an unstranded polyA+ library sequenced on the Illumina HiSeq 2000 and HiSeq 2500 platforms.
* 76-bp paired end reads with a coverage goal of 50M (median achieved was ~82M total reads).
* Raw sequencing data processed with Monorail. Monorail uses STAR to align RNA-seq reads in a spliced fashion to a reference genome, without using any annotation.
* GENCODE 26 transcriptome definition. 

## Proportion of splicing noise

To be able to stablish fair comparisons across different age supergroups, we first had to account for any differences lying within the number of samples from each age cluster.

To account for these differences, we calculated the proportion of splicing noise lying at each distance point and age super-group. 

This proportion was calculated by dividing the total number of novel events located at a particular distance point by the total number of novel junctions available within the window of -60/+60bp distances to the reference intron.


```{r distances_plot_proportion, warning=FALSE}

age_stratification_plot_distances_proportion(df = df_age_groups_novel_tidy, 
                                             age_levels = c("60-79","40-59","20-39" ), 
                                             distance_limit = 60) %>% 
  print()

```



## MSR_Donor

There is a slight increase with age of highly-misspliced introns (peak at MSR_D ~ 1). Digging into the properties of these highly-misspliced introns, they seem to present such a high MSR_D values because they present novel donor junctions that are highly shared among individuals. 

```{r MSRD_Plot, warning=FALSE}

age_stratification_plot_MSRD(df = df_age_groups_intron_tidy,
                             age_levels = c("20-39","40-59","60-79" )) %>% 
  print()

```


## MSR_Donor - Wilcoxon Tests

To statistically test whether overlapping introns across age supergroups presented a change in the distribution of their MSR_D toward smaller values within the 20-39 supergroup than in the 40-59, and also smaller than in 60-79, we used the Wilcoxon test. 

We used the MSR measures to stablish a fair comparison across the age supergroups because these measures were calculated taking across samples from each supergroup. Therefore, both MSR_D and MSR_A measures were normalised figures that contemplate the different number of reads assigned to every intron across samples from each age cluster.

Results from this analysis showed that *the median value corresponding to the distribution of MSR_D values of the overlapping introns across age supergroups significantly increased with age*:

```{r MSR_D_wilcoxon, warning=FALSE}

## Testing using Wilcoxon

df_wilcoxon <- df_age_groups_intron_tidy %>%
  select(ref_junID, sample_type, MSR_D = ref_missplicing_ratio_tissue_ND) %>%
  mutate(MSR_D = MSR_D %>% round(digits = 4)) %>%
  spread(sample_type, MSR_D)

wilcox.test(x = df_wilcoxon$`20-39`,
            y = df_wilcoxon$`40-59`,
            alternative = "less") %>% print()


wilcox.test(x = df_wilcoxon$`20-39`,
            y = df_wilcoxon$`60-79`,
            alternative = "less")%>% print()


wilcox.test(x = df_wilcoxon$`40-59`,
            y = df_wilcoxon$`60-79`,
            alternative = "less")%>% print()

```


## MSR_Acceptor 

There is a slight increase with age of highly-misspliced introns (peak at MSR_A ~ 1). Digging into the properties of these highly-misspliced introns, they seem to present such a high MSR_A values because they present novel acceptor junctions that are highly shared among individuals. 

```{r MSRA_Plot, warning=FALSE}

age_stratification_plot_MSRA(df = df_age_groups_intron_tidy,
                             age_levels = c("20-39","40-59","60-79" )) %>% print()

```

## MSR_Acceptor - Wilcoxon Tests

To statistically test whether overlapping introns across age supergroups presented a change in the distribution of their MSR_A toward smaller values within the 20-39 supergroup than in the 40-59, and also smaller than in 60-79, we used the Wilcoxon test.

Results from this analysis showed that *the median value corresponding to the distribution of MSR_A values of the overlapping introns across age supergroups significantly increased with age*:
  
```{r MSR_A_wilcoxon, warning=FALSE}

df_wilcoxon_MSR_A <- df_age_groups_intron_tidy %>%
  select(ref_junID, sample_type, MSR_A = ref_missplicing_ratio_tissue_NA) %>%
  mutate(MSR_A = MSR_A %>% round(digits = 4)) %>%
  spread(sample_type, MSR_A)


wilcox.test(x = df_wilcoxon_MSR_A$`20-39`,
            y = df_wilcoxon_MSR_A$`40-59`,
            alternative = "less") %>% print()


wilcox.test(x = df_wilcoxon_MSR_A$`20-39`,
            y = df_wilcoxon_MSR_A$`60-79`,
            alternative = "less")%>% print()


wilcox.test(x = df_wilcoxon_MSR_A$`40-59`,
            y = df_wilcoxon_MSR_A$`60-79`,
            alternative = "less")%>% print()

```



# GO Enrichment {.tabset .tabset-fade .tabset-pills}

We obtained the gene enrichment of the genes containing introns whose MSR_D values explicitly increased with age.

The approach followed consisted of:

* Obtain all common introns whose MSR_D value is always smaller in "20-39" than in "40-59" than in "60-79" (i.e introns that are MORE misspliced with age). 
* Obtain their gene name.
* From this data set of genes, remove any overlapping gene containing introns whose MSR_D value is always greater in "20-39" than in "40-59" than in "60-79" (i.e introns that are LESS misspliced as age increases).
* Perform their Go & KEGG Enrichment.
* The custom background used was formed by all genes from all common introns across age super-groups.


## MSR_D increasing with age

```{r GO_MRSD, warning = FALSE, message = F, echo = FALSE}

#####################################
## NEVER MISSPLICED - GENE ENRICHMENT
#####################################

df_MSRD <- df_age_groups_intron_tidy %>%
  select(ref_junID,
         sample_type,
         MSR_D = ref_missplicing_ratio_tissue_ND,
         gene_name) %>%
  mutate(MSR_D = MSR_D %>% round(digits = 4)) %>%
  spread(sample_type, MSR_D)


genes_MSRD_increasing <- df_MSRD %>%
  filter(`20-39` < `40-59`,
         `40-59` < `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name) 


genes_MSRD_dcreasing <- df_MSRD %>%
  filter(`20-39` > `40-59`,
         `40-59` > `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name)



gene_enrichment <- gprofiler2::gost(query = setdiff(genes_MSRD_increasing$gene_name, 
                                                    genes_MSRD_dcreasing$gene_name),
                                    custom_bg = bg_genes,
                                    organism = "hsapiens",
                                    ordered_query = F,
                                    correction_method = "bonferroni",
                                    significant = T)

GO_enrichment <- gene_enrichment$result %>% filter(str_detect(source, pattern = "GO")) %>%
  mutate(go_type = str_sub(source, start = 4, end = 5))

GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                             go_id = GO_enrichment$term_id,
                                                             go_term = GO_enrichment$term_name),
                                     orgdb = "org.Hs.eg.db",
                                     threshold = 0.7,
                                     scores = NULL,
                                     measure = "Wang"
)

df_result <- merge(x = gene_enrichment$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                   y = GO_enrich_reduc,
                   by.x = "term_id",
                   by.y = "go_id",
                   all.x = T)

## JOIN with KEGG results and add pvalues

rbind(df_result %>%
        as_tibble() %>%
        dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
        group_by(parent_term, source) %>%
        mutate(p_value = p_value %>% max) %>%
        ungroup() %>%
        distinct(parent_term, .keep_all = T),
      df_result %>% 
        filter(source == "KEGG")%>%
        dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
        dplyr::rename(parent_term = term_name)) %>%
  arrange(p_value) %>%
  mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
  #filter(source == "KEGG") %>%
  DT::datatable()




```


## MSR_A increasing with age

```{r GO_MRSA, warning = FALSE, message = F, echo = FALSE}

df_MSRA <- df_age_groups_intron_tidy %>%
  dplyr::select(ref_junID,
                sample_type,
                MSR_A = ref_missplicing_ratio_tissue_NA,
                gene_name) %>%
  mutate(MSR_A = MSR_A %>% round(digits = 4)) %>%
  spread(sample_type, MSR_A)


genes_MSRA_increasing <- df_MSRA %>%
  filter(`20-39` < `40-59`,
         `40-59` < `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name)

genes_MSRA_dcreasing <- df_MSRA %>%
  filter(`20-39` > `40-59`,
         `40-59` > `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name)

gene_enrichment <- gprofiler2::gost(query = setdiff(genes_MSRA_increasing$gene_name, 
                                                    genes_MSRA_dcreasing$gene_name),
                                    custom_bg = bg_genes,
                                    organism = "hsapiens",
                                    ordered_query = F,
                                    correction_method = "bonferroni",
                                    significant = T)


GO_enrichment <- gene_enrichment$result %>% filter(str_detect(source, pattern = "GO")) %>%
  mutate(go_type = str_sub(source, start = 4, end = 5))

GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                             go_id = GO_enrichment$term_id,
                                                             go_term = GO_enrichment$term_name),
                                     orgdb = "org.Hs.eg.db",
                                     threshold = 0.7,
                                     scores = NULL,
                                     measure = "Wang"
)

df_result <- merge(x = gene_enrichment$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                   y = GO_enrich_reduc,
                   by.x = "term_id",
                   by.y = "go_id",
                   all.x = T)

## JOIN with KEGG results and add pvalues

rbind(df_result %>%
        as_tibble() %>%
        dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
        group_by(parent_term, source) %>%
        mutate(p_value = p_value %>% max) %>%
        ungroup() %>%
        distinct(parent_term, .keep_all = T),
      df_result %>% 
        filter(source == "KEGG")%>%
        dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
        dplyr::rename(parent_term = term_name)) %>%
  arrange(p_value) %>%
  mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
  #filter(source == "KEGG") %>%
  DT::datatable()


```

<!-- # RBP TPM  {.tabset .tabset-fade .tabset-pills} -->



<!-- ## Covariate correction -->

<!-- TPM values were calculated following the methodology below: -->

<!-- * 1. Calculate the TPM expression of the 150 ENCODE RBPs across samples from GTEX v8 data. The TPM value has been is calculated by: -->
<!--     * 1. Divide the read counts by the length of each gene in kilobases (i.e. RPK) -->
<!--     * 2. Count up all the RPK values in a sample and divide this number by 1,000,000. -->
<!--     * 3. Divide the RPK values by the "per million" scaling factor. -->
<!--     * 4. TPMs were then log10 and normalised. -->

<!-- * 2. Get the best covariates to correct for: -->
<!--     * 1. TPM values were covariate corrected following the approach proposed by Fairbrother-Browne, A., Ali, A.T., Reynolds, R.H. et al. Mitochondrial-nuclear cross-talk in the human brain is modulated by cell type and perturbed in neurodegenerative disease. Commun Biol 4, 1262 (2021). https://doi.org/10.1038/s42003-021-02792-w. -->
<!--     * 2. The covariate selected were: RIN, four batch variables (type of nucleic acid isolation batch, nucleic acid isolation batch ID, genotype or expression batch ID, date of genotype or expression batch), centre, age, gender and cause of death.  -->


<!-- ## Mean TPM   -->

<!-- The heat map below contains the mean TPM values calculated across samples from the same cluster per RBP. -->

<!-- The mean function was chosen here as the TPM values had been previously normalised using a log10 function. -->

<!-- There are some RBPs which present much higher mean covariate-corrected TPM values in 20-39 samples than in 40-59 and 60-79 samples (please, notice the darker red colours). -->

<!-- ```{r RBP_mean_tpm, warning=FALSE} -->

<!--   tpm_corrected <- read.csv(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/", -->
<!--                                           project_id, "/results/pipeline3/rbp/tpm_residuals.csv"), -->
<!--                             header = T) -->
<!--   # tpm_corrected[1,] -->
<!--   # age_samples_clusters_tidy[1,] -->



<!--   tpm_20_39 <- tpm_corrected %>% -->
<!--     filter(X %in% (age_samples_clusters_tidy %>% -->
<!--                      filter(age_group == "20-39") %>% pull(individual))) %>% -->
<!--     gather(key = "RBP", value = "TPM", -X ) %>% -->
<!--     mutate(age_text = paste("20-39_", X)) %>% -->
<!--     mutate(age = "20-39") %>% as_tibble() -->


<!--   tpm_40_59 <-tpm_corrected %>% -->
<!--     filter(X %in% (age_samples_clusters_tidy %>% -->
<!--                      filter(age_group == "40-59") %>% pull(individual)))  %>% -->
<!--     gather(key = "RBP", value = "TPM", -X ) %>% -->
<!--     mutate(age_text = paste("40-59_", X)) %>% -->
<!--     mutate(age = "40-59") %>% as_tibble() -->

<!--   tpm_60_79 <- tpm_corrected %>% -->
<!--     filter(X %in% (age_samples_clusters_tidy %>% -->
<!--                      filter(age_group == "60-79") %>% pull(individual))) %>% -->
<!--     gather(key = "RBP", value = "TPM", -X ) %>% -->
<!--     mutate(age_text = paste("60-79_", X)) %>% -->
<!--     mutate(age = "60-79") %>% as_tibble() -->

<!--   gattered_matrix <- do.call("rbind", list(tpm_20_39, tpm_40_59, tpm_60_79)) %>% -->
<!--     dplyr::rename(sample = X) %>% as_tibble() %>% -->
<!--     group_by(age, RBP) %>% -->
<!--     mutate(mean_tpm = (TPM %>% mean())) %>% -->
<!--     distinct(mean_tpm, .keep_all = T)  %>% -->
<!--     ungroup() %>% -->
<!--     arrange(age , mean_tpm) %>% -->
<!--     mutate(RBP = fct_inorder(RBP)) -->



<!--   ggplot(data = gattered_matrix,  -->
<!--          aes(x=age, y=RBP, fill = mean_tpm)) + -->
<!--     geom_tile() +  -->
<!--     scale_fill_gradient(low = "green", high = "red") + -->
<!--     ggtitle("Mean RBP level of expression across samples\nfrom each age cluster.\nTPM values have been covariate corrected.")+ -->
<!--     ylab("RBP") + -->
<!--     xlab("age cluster") + -->
<!--     theme(axis.text.y = element_blank()) %>%  -->
<!--     return() -->
<!--   #intersect(tpm_20_39$X, tpm_60_79$X) -->



<!-- ``` -->

<!-- ## Stats -->

<!-- ```{r RBP_stats, warning = FALSE, message = F, echo = FALSE} -->

<!-- spread_matrix <- gattered_matrix %>% -->
<!--   dplyr::select(RBP, age, mean_tpm) %>% -->
<!--   spread(key = age, value = mean_tpm) -->

<!-- ref_tidy <- ref %>%  -->
<!--   as_tibble() %>%  -->
<!--   distinct(gene_id, .keep_all = T) %>% -->
<!--   filter(gene_id %in% spread_matrix$RBP) %>% -->
<!--   dplyr::select(gene_id, gene_name) %>%  -->
<!--   as.data.table() -->

<!-- spread_matrix <- merge(x = spread_matrix, -->
<!--                             y = ref_tidy, -->
<!--                             by.x = "RBP", -->
<!--                             by.y = "gene_id") -->



<!-- print(paste0("Total number of RBPs studied: ", spread_matrix %>% nrow())) -->

<!-- print(paste0("RBPs whose TPM decrease in each age group: ", spread_matrix %>% filter(`20-39` >  `40-59`, `40-59` > `60-79`) %>% nrow())) -->
<!-- print(paste0("RBPs whose TPM decrease from '20-39' compared to the other two age groups: ", spread_matrix %>% filter(`20-39` > `40-59`, `20-39` > `60-79`) %>% nrow())) -->

<!-- print(paste0("RBPs whose TPM increase in each age group: ", spread_matrix %>% filter(`20-39` <  `40-59`, `40-59` < `60-79`) %>% nrow())) -->

<!-- print(paste0("RBPs whose TPM increase from '20-39' compared to the other two age groups: ", spread_matrix %>% filter(`20-39` < `40-59`, `20-39` < `60-79`) %>% nrow())) -->

<!-- # spread_matrix %>% filter(`20-39` <=  `40-59`) -->
<!-- # spread_matrix %>% filter(`20-39` <= `40-59`) -->
<!-- # spread_matrix %>% filter(`20-39` >= `40-59`, `40-59` >= `60-79`) -->
<!-- ``` -->

<!-- ## RBP Go Enrichment -->

<!-- I obtained the gene enrichment of the 272 RBPs whose TPM values explicitly decreased with age from "20-39" to "40-59". -->

<!-- The approach followed consisted of: -->

<!-- * Obtain all RBPs whose covariate-corrected TPM values explicitly decreased with age from "20-39" to "40-59". -->
<!-- * Obtain their gene name. -->
<!-- * Perform their Go & KEGG Enrichment. -->
<!-- * No custom background used. When using the dataset of 272 RBPs as bg, the query did not return any significant results.  -->

<!-- ```{r RBP_go_enrichment, warning = FALSE, message = F, echo = FALSE} -->

<!-- ##################################### -->
<!-- ## NEVER MISSPLICED - GENE ENRICHMENT -->
<!-- ##################################### -->


<!-- spread_matrix_tidy <- spread_matrix %>% -->
<!--   filter(`20-39` > `40-59`, -->
<!--          `20-39` > `60-79`) %>%  -->
<!--   as.data.table() -->



<!-- gene_enrichment <- gprofiler2::gost(query = spread_matrix_tidy$gene_name, -->
<!--                                     #custom_bg = spread_matrix$gene_name, -->
<!--                                     organism = "hsapiens", -->
<!--                                     ordered_query = F, -->
<!--                                     correction_method = "bonferroni", -->
<!--                                     significant = T) -->

<!-- GO_enrichment <- gene_enrichment$result %>%  -->
<!--   filter(str_detect(source, pattern = "GO")) %>% -->
<!--   mutate(go_type = str_sub(source, start = 4, end = 5)) -->

<!-- GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type, -->
<!--                                                              go_id = GO_enrichment$term_id, -->
<!--                                                              go_term = GO_enrichment$term_name), -->
<!--                                      orgdb = "org.Hs.eg.db", -->
<!--                                      #threshold = 0.7, -->
<!--                                      scores = NULL, -->
<!--                                      measure = "Wang" -->
<!-- ) -->

<!-- df_result <- merge(x = gene_enrichment$result %>%  -->
<!--                      filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")), -->
<!--                    y = GO_enrich_reduc, -->
<!--                    by.x = "term_id", -->
<!--                    by.y = "go_id", -->
<!--                    all.x = T) -->

<!-- rbind(df_result %>% -->
<!--         as_tibble() %>% -->
<!--         dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>% -->
<!--         group_by(parent_term, source) %>% -->
<!--         mutate(p_value = p_value %>% max) %>% -->
<!--         ungroup() %>% -->
<!--         distinct(parent_term, .keep_all = T) %>% -->
<!--         drop_na(),  -->
<!--       df_result %>% -->
<!--         as_tibble() %>% -->
<!--         dplyr::select(parent_term = term_name, p_value, query_size, intersection_size, source) %>% -->
<!--         filter(source == "KEGG") %>% -->
<!--         group_by(parent_term, source) %>% -->
<!--         mutate(p_value = p_value %>% max) %>% -->
<!--         ungroup() %>% -->
<!--         distinct(parent_term, .keep_all = T))%>% -->
<!--   arrange(p_value) %>% -->
<!--   mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>% -->
<!--   #filter(source == "KEGG") %>% -->
<!--   DT::datatable() -->

<!-- ``` -->

<!-- ## U2AF1 -->

<!-- Focusing now in a core spliceosomal RBP responsible for targeting the 3'ss signal, I ran a Wilcoxon test to test whether the set of TPM values across samples from 20-39 presented a skewed distribution towards bigger values than in 40-59 and 60-79 samples. -->

<!-- ```{r U2AF1_tpm, warning=FALSE} -->

<!-- U2AF1_20_39 <- tpm_20_39 %>% filter(RBP == "ENSG00000160201") %>% pull(TPM) -->
<!-- U2AF1_40_59 <- tpm_40_59 %>% filter(RBP == "ENSG00000160201") %>% pull(TPM) -->
<!-- U2AF1_60_79 <- tpm_60_79 %>% filter(RBP == "ENSG00000160201") %>% pull(TPM) -->

<!-- wilcox.test(x = U2AF1_20_39,  -->
<!--             y = U2AF1_40_59,  -->
<!--             correct = T, -->
<!--             alternative = "greater") %>% print() -->

<!-- wilcox.test(x = U2AF1_20_39,  -->
<!--             y = U2AF1_60_79,  -->
<!--             correct = T, -->
<!--             alternative = "greater") %>% print() -->


<!-- wilcox.test(x = U2AF1_40_59,  -->
<!--             y = U2AF1_60_79, -->
<!--             correct = T, -->
<!--             alternative = "greater") %>% print() -->


<!-- ``` -->

<!-- ## U2AF2 -->

<!-- ```{r U2AF2_tpm, warning=FALSE} -->

<!-- U2AF2_20_39 <- tpm_20_39 %>% filter(RBP == "ENSG00000063244") %>% pull(TPM) -->
<!-- U2AF2_40_59 <- tpm_40_59 %>% filter(RBP == "ENSG00000063244") %>% pull(TPM) -->
<!-- U2AF2_60_79 <- tpm_60_79 %>% filter(RBP == "ENSG00000063244") %>% pull(TPM) -->

<!-- wilcox.test(x = U2AF2_20_39,  -->
<!--             y = U2AF2_40_59,  -->
<!--             correct = T, -->
<!--             alternative = "greater") %>% print() -->

<!-- wilcox.test(x = U2AF2_20_39,  -->
<!--             y = U2AF2_60_79,  -->
<!--             correct = T, -->
<!--             alternative = "greater") %>% print() -->


<!-- wilcox.test(x = U2AF2_40_59,  -->
<!--             y = U2AF2_60_79, -->
<!--             correct = T, -->
<!--             alternative = "greater") %>% print() -->


<!-- ``` -->


<!-- # RBP Linear Models  {.tabset .tabset-fade .tabset-pills} -->

<!-- Each RBP studied had a distribution of TPM values obtained from all samples belonging to the current GTEx tissue (one TPM value assigned to each sample). Each sample has assigned a set of covariates. -->

<!-- We run a **linear model per RBP** to predict its uncorrected TPM value across the samples. We used the sample covariates (RIN, four batch variables (type of nucleic acid isolation batch, nucleic acid isolation batch ID, genotype or expression batch ID, date of genotype or expression batch), centre, age, gender and cause of death) as predictors. -->

<!-- **Once we had run the 272 linear modes, one per RBP studied, we filtered them to obtain only the significant (pval < 0.05) beta values corresponding to the categorical variable age.** -->

<!-- The table below therefore **ONLY contains the RBPs that obtained a significant beta value** assigned to the "age" feature in the prediction of their uncorrected TPM value in the linear models.  -->

<!-- 0 values correspond to non-significant beta values. -->

<!-- ```{r RBP_lm, echo = FALSE} -->

<!-- tpm_lm <- read.csv(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/", project_id, -->
<!--                                  "/results/pipeline3/rbp/tpm_lm.csv")) %>% -->
<!--   filter(pval < 0.05) %>%  -->
<!--   dplyr::select(-pval) -->

<!-- tpm_lm %>% -->
<!--   spread(key = covariate, value = "Estimate") %>% -->
<!--   replace(is.na(.),0) %>% -->
<!--   as_tibble() %>% -->
<!--   DT::datatable() -->

<!-- ``` -->


<!-- ## TPM decreasing with age -->

<!-- To ease the interpretation of the table above, we filtered it to obtain only the RBPs that showed a decreasing uncorrected TPM value with age. In other words, we obtained the RBPs that obtained a negative estimate value for age in the linear models: -->

<!-- ```{r rbp_lm_decreasing, echo = FALSE} -->

<!-- RBP_decrease <- tpm_lm %>% -->
<!--     filter(Estimate < 0) %>% -->
<!--     arrange(desc(abs(Estimate))) %>% -->
<!--   distinct(hgnc_symbol) %>% -->
<!--   pull -->

<!-- RBP_decrease %>% sort %>% print() -->

<!-- ``` -->

<!-- Then, in order to try to match results with the previous analysis in which we used the covariate-corrected TPM value to analyse changes between age groups, we **obtained the overlap of RBPs from the previous analysis and the dataset of RBPs whose covariate-corrected TPM decreased from the '20-39' to the other two age groups.** Below are the overlapping RBPs between both analyses: -->

<!-- ```{r rbp_lm_decreasing_overlapped, echo = FALSE} -->

<!-- intersect(x = RBP_decrease, -->
<!--           y = spread_matrix %>% filter(`20-39` > `40-59`, `20-39` > `60-79`) %>% distinct(gene_name) %>% pull()) %>% sort -->

<!-- ``` -->

<!-- The GO Enrichment below of these decreasing-expression-with-age RBPs (using as background the dataset of all 250 RBPs) did not return significant terms. -->


<!-- ## TPM increasing with age -->

<!-- Similarly, from the dataset of RBPs obtaining a significant age result in the linear models, the following ones presented increasing (i.e. positive estimate result) in the prediction of their uncorrected TPM values with age: -->

<!-- ```{r rbp_lm_increasing, echo = FALSE} -->

<!-- RBP_increase <- tpm_lm %>% -->
<!--     filter(Estimate > 0) %>% -->
<!--     arrange(desc(abs(Estimate))) %>% -->
<!--   distinct(hgnc_symbol) %>% -->
<!--   pull -->

<!-- RBP_increase %>% sort() %>% print() -->

<!-- ``` -->


<!-- From them, the following RBPs overlapped between the two analyses (RBPs obtaining a positive significant beta value in the linear models with the RBPs whose covariate-corrected TPM increased with age, from the '20-39' to the other two age groups): -->


<!-- ```{r rbp_lm_increasing_overlapped, echo = FALSE} -->

<!-- intersect(x = RBP_increase, -->
<!--           y = spread_matrix %>%  filter(`20-39` < `40-59`, `20-39` < `60-79`) %>% distinct(gene_name) %>% pull()) %>% sort() -->

<!-- ``` -->


<!-- Their GO Enrichment below (no background used as when the dataset of all RBPs was used as a bg, no significant results were returned): -->

<!-- ```{r rbp_lm_increasing_go, echo = FALSE} -->

<!-- if (RBP_increase %>% length() > 0) { -->
<!--   ## Get RBPs that increase expression with age -->
<!--   gene_enrichment <- gprofiler2::gost(query = RBP_increase %>% unique(), -->
<!--                                       #custom_bg = RBP_bg, -->
<!--                                       organism = "hsapiens", -->
<!--                                       ordered_query = T, -->
<!--                                       correction_method = "bonferroni", -->
<!--                                       significant = T) -->

<!--   GO_enrichment <- gene_enrichment$result %>%  -->
<!--     filter(str_detect(source, pattern = "GO")) %>% -->
<!--     mutate(go_type = str_sub(source, start = 4, end = 5)) -->

<!--   GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type, -->
<!--                                                                go_id = GO_enrichment$term_id, -->
<!--                                                                go_term = GO_enrichment$term_name), -->
<!--                                        orgdb = "org.Hs.eg.db", -->
<!--                                        #threshold = 0.7, -->
<!--                                        scores = NULL, -->
<!--                                        measure = "Wang") -->

<!--   df_result <- merge(x = gene_enrichment$result %>%  -->
<!--                        filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")), -->
<!--                      y = GO_enrich_reduc, -->
<!--                      by.x = "term_id", -->
<!--                      by.y = "go_id", -->
<!--                      all.x = T) -->

<!--   rbind(df_result %>% -->
<!--           as_tibble() %>% -->
<!--           dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>% -->
<!--           group_by(parent_term, source) %>% -->
<!--           mutate(p_value = p_value %>% max) %>% -->
<!--           ungroup() %>% -->
<!--           distinct(parent_term, .keep_all = T) %>% -->
<!--           drop_na(),  -->
<!--         df_result %>% -->
<!--           as_tibble() %>% -->
<!--           dplyr::select(parent_term = term_name, p_value, query_size, intersection_size, source) %>% -->
<!--           filter(source == "KEGG") %>% -->
<!--           group_by(parent_term, source) %>% -->
<!--           mutate(p_value = p_value %>% max) %>% -->
<!--           ungroup() %>% -->
<!--           distinct(parent_term, .keep_all = T))%>% -->
<!--     arrange(p_value) %>% -->
<!--     mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>% -->
<!--     DT::datatable() -->
<!-- } else { -->
<!--   print("There are no RBPs showing increasing levels of expression with age.") -->
<!-- } -->

<!-- ``` -->
