---
title: "Age Stratification - MUSCLE"
author: "Sonia Garcia-Ruiz"
date: "27/03/2022"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
    #number_sections: true
    theme: flatly
    highlight: tango
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(rutils)

project_id <- "MUSCLE"

source(paste0("/home/sruiz/PROJECTS/splicing-project-recount3/pipeline4-2_age_stratification.R"))

ref <- rtracklayer::import(con = "/data/references/ensembl/gtf/v105/Homo_sapiens.GRCh38.105.chr.gtf")

age_samples_clusters_tidy <- age_stratification_init_data(project_id)
age_supergroups <- age_samples_clusters_tidy$age_group %>% unique()


folder_root <- paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/", project_id,
                      "/results/pipeline3/missplicing-ratio/age/")

## Load the IDBs
df_age_groups_intron <- map_df(age_supergroups, function(age_group) {
  
  # age_group <- age_supergroups[1]
  readRDS(file = paste0(folder_root, "/", age_group, "/", age_group, "_db_introns.rds")) %>%
    #mutate(ref_junID = ref_junID %>% as.integer()) %>%
    mutate(sample_type = age_group) %>%
    return()
  
})

df_age_groups_novel <- map_df(age_supergroups, function(age_group) {
  
  #print(paste0(Sys.time(), " - loading IDB for '", age_group, "' samples ..."))
  readRDS(file = paste0(folder_root, "/", age_group, "/", age_group, "_db_novel.rds")) %>%
    #mutate(ref_junID = ref_junID %>% as.integer()) %>%
    #mutate(novel_junID = novel_junID %>% as.integer()) %>%
    mutate(sample_type = age_group) %>%
    return()
  
}) 

## Get the junctions that are common across age supergroups

common_introns_misspliced <- df_age_groups_novel %>%
  group_by(sample_type) %>%
  distinct(ref_junID, .keep_all = T) %>%
  ungroup() %>%
  dplyr::count(ref_junID) %>%
  filter(n == age_supergroups %>% length()) %>%
  pull(ref_junID)

common_introns <- df_age_groups_intron %>%
  group_by(sample_type) %>%
  distinct(ref_junID, .keep_all = T) %>%
  ungroup() %>%
  dplyr::count(ref_junID) %>%
  filter(n == age_supergroups %>% length()) %>%
  pull(ref_junID)

# setdiff(common_introns_misspliced,common_introns)
# setdiff(common_introns,common_introns_misspliced)
# common_introns_misspliced %>% length()
# common_introns %>% length()
# intersect(common_introns_misspliced,common_introns) %>% length()

## Filter by the common junctions
df_age_groups_intron_tidy <- merge(x = df_age_groups_intron %>% data.table::as.data.table(),
                                   y = data.table::data.table(ref_junID = common_introns_misspliced),
                                   by = "ref_junID",
                                   all.y = T)

saveRDS(object = df_age_groups_intron_tidy,
        file = paste0(folder_root, "/common_introns_age.rds"))

df_age_groups_novel_tidy <- merge(x = df_age_groups_novel %>% data.table::as.data.table(),
                                  y = data.table::data.table(ref_junID = common_introns_misspliced),
                                  by = "ref_junID",
                                  all.y = T)



bg_genes <- df_age_groups_novel_tidy %>% distinct(gene_name) %>% pull() %>% unlist %>% unique()
# bg_genes2 <- df_age_groups_intron_tidy %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
# 
# 
# bg_genes %>% length()
# bg_genes2 %>% length()
# identical(bg_genes,bg_genes2)

```

# Data used

This document contains the age analysis performed using RNA-seq data from the GTEx v8 project and processed by [recount3](https://jhubiostatistics.shinyapps.io/recount3-study-explorer/).

Samples were grouped together within the following age clusters:

* 20-39: contains samples from the age groups "20-29" and "30-39"
* 40-59: contains samples from the age groups "40-49" and "50-59"
* 60-79: contains samples from the age groups "60-69" and "70-79".


# Stats table {.tabset .tabset-fade .tabset-pills}

Samples have been displayed after being grouped by different columns such as 'age', 'sex', 'RIN' and 'body region'.

## Across tissues

```{r mean_values}

age_samples_clusters_tidy %>%
  select(-individual, -region, -age, -sample_recount_id, -region) %>%
  group_by(age_group) %>%
  mutate(median_rin = rin %>% median())%>%
  mutate(mean_read_depth = round(mapped_read_count %>% mean())) %>%
  select(-rin,-mapped_read_count) %>%
  mutate(N_samples = n()) %>%
  ungroup() %>%
  #group_by(region, age_group) %>%
  #mutate(N_samples_tissue = n()) %>%
  #ungroup() %>%
  group_by(age_group, sex) %>%
  mutate(sex_ratio = round(n()/N_samples, digits = 2)) %>%
  distinct(age_group, .keep_all = T) %>%
  spread(key = sex, value = sex_ratio) %>%
  dplyr::rename(ratio_females = female,
                ratio_males = male) %>%
  relocate(N_samples, .after = age_group) %>%
  DT::datatable(rownames = F)
  
```


Column description:

* **N_samples:** total number of samples clustered within the age super-group.
* **median_rin:** this is the median RIN number of all samples clustered within the age group. 
* **mean_read_depth**: this is the mean 'mapped_read_depth' of all samples clustered within the age group. 
* **ratio_females:** this is the ratio of female samples available within the age group.
* **ratio_males:** this is the ratio of male samples available within the age group.


## Per tissue

```{r brain_region}

age_samples_clusters_tidy %>% as_tibble() %>%
  select(-individual, -age, -sex,-sample_recount_id) %>%
  group_by(age_group) %>%
  mutate(median_rin = rin %>% median()) %>%
  mutate(mean_read_depth = round(mapped_read_count %>% mean())) %>%
  select(-rin,-mapped_read_count) %>%
  mutate(N_samples = n()) %>%
  ungroup() %>%
    
  group_by(region, age_group) %>%
  mutate(N_samples_tissue = n()) %>%
  ungroup() %>%

  # group_by(age_group, sex) %>%
  # mutate(sex_ratio = round(x = n()/N_samples,
  #                          digits = 2)) %>%
  # ungroup() %>%
  
  group_by(region, age_group) %>%
  distinct(age_group, .keep_all = T) %>%
  #spread(key = sex, value = sex_ratio) %>%
  dplyr::arrange(region) %>%
    #ungroup() %>%
    #group_by(age_group) %>%
    #mutate(mean_samples_across_tissues = N_samples_tissue %>% mean) %>%
    #mutate(median_samples_across_tissues = N_samples_tissue %>% median) %>%
  spread(key = region, value = N_samples_tissue) %>% 
    #select(-region, -N_samples_tissue) %>%
  distinct(age_group, .keep_all = T) %>% 
  #spread(key = sex, value = sex_ratio)  %>%
  as.data.frame()  %>%
  # dplyr::rename(ratio_females = female,
  #               ratio_males = male) %>%
  relocate(N_samples, .after = age_group) %>%
  DT::datatable(rownames = F)
  
```


Column description:

* **N_samples:** total number of samples clustered within the age group.
* **median_rin:** this is the median RIN number of all samples clustered within the age group. 
* **mean_read_depth**: this is the mean 'mapped_read_depth' of all samples clustered within the age group. 
* **tissue**: This is the exact number of samples available per body region clustered within each age group.


# Plots {.tabset .tabset-fade .tabset-pills}

## Distances

<!-- In this graph it has only been included **116,847** reference introns that were common across all three age supergroups. -->
<!-- The plot below shows the distances from all novel junction located within a window of -60/+60 bp to their reference introns.  -->

In this graph it has only been included common reference introns across all three age supergroups.
The plot below shows the distances from all novel junction located within a window of -30/+30 bp from their reference introns.

Interestingly, the distances peak at the donor exon location is not there anymore if we compare these results with the ones obtained using GTEx v6 data. This is the same type of distances result that we obtained using data from the ENCODE RBP knockdown projects. 


```{r distances_plot, warning=FALSE}

age_stratification_plot_distances(df = df_age_groups_novel_tidy, 
                                  age_levels = c("60-79","40-59","20-39" ), 
                                  distance_limit = 60) %>% print()

```

ENCODE Pipeline overview:

* Illumina TruSeq protocol to create a polyA selected strand specific library.
* RNA populations ~200bp 
* Mapping of the reads to the genome using STAR.
* Gene annotation files using GENCODE

Recount3/GTEx v8 pipeline overview:

* The Illumina TruSeq protocol was used to create an unstranded polyA+ library sequenced on the Illumina HiSeq 2000 and HiSeq 2500 platforms.
* 76-bp paired end reads with a coverage goal of 50M (median achieved was ~82M total reads).
* Raw sequencing data processed with Monorail. Monorail uses STAR to align RNA-seq reads in a spliced fashion to a reference genome, without using any annotation.
* GENCODE 26 transcriptome definition. 

## Proportion of splicing noise

To be able to stablish fair comparisons across different age supergroups, we first had to account for any differences lying within the number of samples from each age cluster.

To account for these differences, we calculated the proportion of splicing noise lying at each distance point and age super-group. 

This proportion was calculated by dividing the total number of novel events located at a particular distance point by the total number of novel junctions available within the window of -60/+60bp distances to the reference intron.


```{r distances_plot_proportion, warning=FALSE}

age_stratification_plot_distances_proportion(df = df_age_groups_novel_tidy, 
                                             age_levels = c("60-79","40-59","20-39" ), 
                                             distance_limit = 60) %>% 
  print()

```



## MSR_Donor

There is a slight increase with age of highly-misspliced introns (peak at MSR_D ~ 1). Digging into the properties of these highly-misspliced introns, they seem to present such a high MSR_D values because they present novel donor junctions that are highly shared among individuals. 

```{r MSRD_Plot, warning=FALSE}

age_stratification_plot_MSRD(df = df_age_groups_intron_tidy,
                             age_levels = c("20-39","40-59","60-79" )) %>% 
  print()

```


## MSR_Donor COMMON

The graph below contains the MSR_D values of common introns across age groups and BRAIN, MUSCLE and BLOOD tissues. 

```{r MSRD_Plot_common, warning=FALSE}

brain_df_age_groups_intron_tidy <- readRDS("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/BRAIN/results/pipeline3/missplicing-ratio/age/common_introns_age.rds")
muscle_df_age_groups_intron_tidy <- readRDS("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/MUSCLE/results/pipeline3/missplicing-ratio/age/common_introns_age.rds")
blood_df_age_groups_intron_tidy <- readRDS("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/BLOOD/results/pipeline3/missplicing-ratio/age/common_introns_age.rds")

common_ref <- Reduce(intersect, list(brain_df_age_groups_intron_tidy$ref_junID,
                                     muscle_df_age_groups_intron_tidy$ref_junID,
                                     blood_df_age_groups_intron_tidy$ref_junID))



age_stratification_plot_MSRD(df = df_age_groups_intron_tidy %>%
                               filter(ref_junID %in% common_ref),
                             age_levels = c("20-39","40-59","60-79" )) %>% print()

```
## MSR_Donor - Wilcoxon Tests

To statistically test whether overlapping introns across age supergroups presented a change in the distribution of their MSR_D toward smaller values within the 20-39 supergroup than in the 40-59, and also smaller than in 60-79, we used the Wilcoxon test. 

We used the MSR measures to stablish a fair comparison across the age supergroups because these measures were calculated taking across samples from each supergroup. Therefore, both MSR_D and MSR_A measures were normalised figures that contemplate the different number of reads assigned to every intron across samples from each age cluster.

Results from this analysis showed that *the median value corresponding to the distribution of MSR_D values of the overlapping introns across age supergroups significantly increased with age*:

```{r MSR_D_wilcoxon, warning=FALSE}

## Testing using Wilcoxon

df_wilcoxon <- df_age_groups_intron_tidy %>%
  select(ref_junID, sample_type, MSR_D = ref_missplicing_ratio_tissue_ND) %>%
  mutate(MSR_D = MSR_D %>% round(digits = 4)) %>%
  spread(sample_type, MSR_D)

wilcox.test(x = df_wilcoxon$`20-39`,
            y = df_wilcoxon$`40-59`,
            alternative = "less") %>% print()


wilcox.test(x = df_wilcoxon$`20-39`,
            y = df_wilcoxon$`60-79`,
            alternative = "less")%>% print()


wilcox.test(x = df_wilcoxon$`40-59`,
            y = df_wilcoxon$`60-79`,
            alternative = "less")%>% print()

```


## MSR_Acceptor 

There is a slight increase with age of highly-misspliced introns (peak at MSR_A ~ 1). Digging into the properties of these highly-misspliced introns, they seem to present such a high MSR_A values because they present novel acceptor junctions that are highly shared among individuals. 

```{r MSRA_Plot, warning=FALSE}

age_stratification_plot_MSRA(df = df_age_groups_intron_tidy,
                             age_levels = c("20-39","40-59","60-79" )) %>% print()

```

## MSR_Acceptor COMMON

The graph below contains the MSR_A values of common introns across age groups and BRAIN, MUSCLE and BLOOD tissues. 

```{r MSRA_Plot_common, warning=FALSE}


age_stratification_plot_MSRA(df = df_age_groups_intron_tidy %>%
                               filter(ref_junID %in% common_ref),
                             age_levels = c("20-39","40-59","60-79" )) %>% print()

```


## MSR_Acceptor - Wilcoxon Tests

To statistically test whether overlapping introns across age supergroups presented a change in the distribution of their MSR_A toward smaller values within the 20-39 supergroup than in the 40-59, and also smaller than in 60-79, we used the Wilcoxon test.

Results from this analysis showed that *the median value corresponding to the distribution of MSR_A values of the overlapping introns across age supergroups significantly increased with age*:
  
```{r MSR_A_wilcoxon, warning=FALSE}

df_wilcoxon_MSR_A <- df_age_groups_intron_tidy %>%
  select(ref_junID, sample_type, MSR_A = ref_missplicing_ratio_tissue_NA) %>%
  mutate(MSR_A = MSR_A %>% round(digits = 4)) %>%
  spread(sample_type, MSR_A)


wilcox.test(x = df_wilcoxon_MSR_A$`20-39`,
            y = df_wilcoxon_MSR_A$`40-59`,
            alternative = "less") %>% print()


wilcox.test(x = df_wilcoxon_MSR_A$`20-39`,
            y = df_wilcoxon_MSR_A$`60-79`,
            alternative = "less")%>% print()


wilcox.test(x = df_wilcoxon_MSR_A$`40-59`,
            y = df_wilcoxon_MSR_A$`60-79`,
            alternative = "less")%>% print()

```



# GO Enrichment {.tabset .tabset-fade .tabset-pills}

We obtained the gene enrichment of the genes containing introns whose MSR_D values explicitly increased with age.

The approach followed consisted of:

* Obtain all common introns whose MSR_D value is always smaller in "20-39" than in "40-59" than in "60-79" (i.e introns that are MORE misspliced with age). 
* Obtain their gene name.
* From this data set of genes, remove any overlapping gene containing introns whose MSR_D value is always greater in "20-39" than in "40-59" than in "60-79" (i.e introns that are LESS misspliced as age increases).
* Perform their Go & KEGG Enrichment.
* The custom background used was formed by all genes from all common introns across age super-groups.


## MSR_D increasing with age

```{r GO_MRSD, warning = FALSE, message = F, echo = FALSE}

#####################################
## NEVER MISSPLICED - GENE ENRICHMENT
#####################################

df_MSRD <- df_age_groups_intron_tidy %>%
                               filter(ref_junID %in% common_ref) %>%
  select(ref_junID,
         sample_type,
         MSR_D = ref_missplicing_ratio_tissue_ND,
         gene_name) %>%
  mutate(MSR_D = MSR_D %>% round(digits = 4)) %>%
  spread(sample_type, MSR_D)


genes_MSRD_increasing <- df_MSRD %>%
  filter(`20-39` < `40-59`,
         `40-59` < `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name) 


genes_MSRD_dcreasing <- df_MSRD %>%
  filter(`20-39` > `40-59`,
         `40-59` > `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name)



gene_enrichment <- gprofiler2::gost(query = setdiff(genes_MSRD_increasing$gene_name, 
                                                    genes_MSRD_dcreasing$gene_name),
                                    custom_bg = bg_genes,
                                    organism = "hsapiens",
                                    ordered_query = F,
                                    correction_method = "bonferroni",
                                    significant = T)

GO_enrichment <- gene_enrichment$result %>% filter(str_detect(source, pattern = "GO")) %>%
  mutate(go_type = str_sub(source, start = 4, end = 5))

GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                             go_id = GO_enrichment$term_id,
                                                             go_term = GO_enrichment$term_name),
                                     orgdb = "org.Hs.eg.db",
                                     threshold = 0.7,
                                     scores = NULL,
                                     measure = "Wang"
)

df_result <- merge(x = gene_enrichment$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                   y = GO_enrich_reduc,
                   by.x = "term_id",
                   by.y = "go_id",
                   all.x = T)

## JOIN with KEGG results and add pvalues

rbind(df_result %>%
        as_tibble() %>%
        dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
        group_by(parent_term, source) %>%
        mutate(p_value = p_value %>% max) %>%
        ungroup() %>%
        distinct(parent_term, .keep_all = T),
      df_result %>% 
        filter(source == "KEGG")%>%
        dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
        dplyr::rename(parent_term = term_name)) %>%
  arrange(p_value) %>%
  mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
  #filter(source == "KEGG") %>%
  DT::datatable()




```


## MSR_A increasing with age

```{r GO_MRSA, warning = FALSE, message = F, echo = FALSE}

df_MSRA <- df_age_groups_intron_tidy  %>%
                               filter(ref_junID %in% common_ref) %>%
  dplyr::select(ref_junID,
                sample_type,
                MSR_A = ref_missplicing_ratio_tissue_NA,
                gene_name) %>%
  mutate(MSR_A = MSR_A %>% round(digits = 4)) %>%
  spread(sample_type, MSR_A)


genes_MSRA_increasing <- df_MSRA %>%
  filter(`20-39` < `40-59`,
         `40-59` < `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name)

genes_MSRA_dcreasing <- df_MSRA %>%
  filter(`20-39` > `40-59`,
         `40-59` > `60-79`) %>%
  mutate_if(is.list, simplify_all) %>%    
  unnest(gene_name)

gene_enrichment <- gprofiler2::gost(query = setdiff(genes_MSRA_increasing$gene_name, 
                                                    genes_MSRA_dcreasing$gene_name),
                                    custom_bg = bg_genes,
                                    organism = "hsapiens",
                                    ordered_query = F,
                                    correction_method = "bonferroni",
                                    significant = T)


GO_enrichment <- gene_enrichment$result %>% filter(str_detect(source, pattern = "GO")) %>%
  mutate(go_type = str_sub(source, start = 4, end = 5))

GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                             go_id = GO_enrichment$term_id,
                                                             go_term = GO_enrichment$term_name),
                                     orgdb = "org.Hs.eg.db",
                                     threshold = 0.7,
                                     scores = NULL,
                                     measure = "Wang"
)

df_result <- merge(x = gene_enrichment$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                   y = GO_enrich_reduc,
                   by.x = "term_id",
                   by.y = "go_id",
                   all.x = T)

## JOIN with KEGG results and add pvalues

rbind(df_result %>%
        as_tibble() %>%
        dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
        group_by(parent_term, source) %>%
        mutate(p_value = p_value %>% max) %>%
        ungroup() %>%
        distinct(parent_term, .keep_all = T),
      df_result %>% 
        filter(source == "KEGG")%>%
        dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
        dplyr::rename(parent_term = term_name)) %>%
  arrange(p_value) %>%
  mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
  #filter(source == "KEGG") %>%
  DT::datatable()


```
