---
title: "IDB - GTExv8 All Tissues - Characterising splicing noise"
author: "Sonia Garc√≠a-Ruiz"
date: "19/05/2022"
output: 
  html_document:
    code_folding: hide
    toc: true
    #toc_depth: 2
    #toc_float:
    #  collapsed: false
    #  smooth_scroll: false
    #number_sections: true
    theme: flatly
    highlight: tango
  pdf_document: default
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE,
                      warning = F)

library(tidyverse)
library(SummarizedExperiment)
library(data.table)
library(GenomicRanges)

library(ggpubr)
library(matrixStats)
library(xlsx)
#library(kableExtra)


gtf_version <- 105

get_mode <- function(vector) {
  uniqv <- unique(vector)
  uniqv[which.max(tabulate(match(vector, uniqv)))]
}

# source("/home/sruiz/PROJECTS/splicing-project/pipeline3_gtex.R")
source("/home/sruiz/PROJECTS/splicing-project-recount3/pipeline4-1_analyse_IDB.R")

```

# All GTEx v8 tissues 

Each analysis provides with a summary table containing the results obtained per each GTEx v8 tissue.

## Summary  {.tabset .tabset-fade .tabset-pills}

### Distances

```{r distances_all, echo = FALSE}

all_data <- readRDS(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/summary_distances.rds"))

all_data %>% 
  spread(key = Var1, value = Freq) %>%
  dplyr::rename("Tissue" = tissue,
                "Type" = NovelType) %>% 
  mutate(across(where(is.numeric), ~ round(., 2))) %>%
  DT::datatable() %>% 
  return()

```


### MSR

```{r MSR_all, echo = FALSE}

all_data <- readRDS(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/summary_MSRs.rds"))

all_data %>% 
  dplyr::rename("Tissue" = tissue) %>% 
  spread(key = Var1, value = Freq) %>%
  DT::datatable() %>%
  DT::formatRound(columns = c(5:8), digits = 5) %>%
  return()
  


```



### Linear Models

```{r lm_all, echo = FALSE}

library(shiny)
library(DT)
library(dplyr)

all_data <- readRDS(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/summary_lm.rds"))

all_data %>% 
  dplyr::rename("Tissue" = tissue) %>% 
  #mutate(.data = across(where(is.double()), format(x = .,scientific = T))) %>%
  #scientific( digits = 3) %>%
  #DT::datatable(options = list(c(3:4), format(x = ., scientific = T))) %>%
  DT::datatable(options = list(
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>1000 | data[i]<1){",
        "$('td:eq('+i+')', row).html(data[i].toExponential(1));",
        "}",
        "}",
        "}")
    )
    ) %>%
  return()
  


```

## Distribution of the beta values

To study any tissue differences in the effect that every each genic feature could be presenting during the generation of mis-splicing at both splice sites, we run two independent linear regression models per tissue to predict mis-splicing at donor and at acceptor splice sites, respectively. To reduce any biases in regards to the introns we used during each analysis, we only included in the linear models the introns that were common to all tissues (i.e. presented at least one read in at least one of the samples of each tissue).

The graph below displays the output obtained from each linear regression analysis represented as the distribution of the estimate values generated by every predictor included in the model executed from every tissue analysed. Each box from the graph below therefore contains 40 different values, each of them corresponding to a different GTEx tissue.

Results showed that only the strength of the splicing signals and the conservation scores of the proximal intronic sequences that were surrounding them presented the highest variability across tissues. Thus, the canonical splicing signals and the mean conservation scores assigned to the proximal intronic sequences are of variable importance in predicting mis-splicing across different human tissues.

**Since the splicing strength and conservation scores assigned to each intron are identical across tissues, mis-splicing cannot be due to changes in the sequence of introns. Instead, this result is consistent with the hypothesis that mis-splicing can be the consequence of nonoptimal levels of expression of the RBPs responsible for recognising the canonical splicing signals that are essential for correct splicing.** 




```{r variance_common,  echo = FALSE}

plot_estimate_variance(brain_tissues = F) %>% print()

```




## Gene Enrichment - Unique genes {.tabset .tabset-fade .tabset-pills}

GO Enrichment analyses of the genes from COMMON reference introns across all GTEx v8 tissues.


### Stats 

Per each category of introns (i.e. both, donor, acceptor and never) we obtained the genes that were unique to each category.

This means that any overlapping gene across different intron categories was discarded and not included in the following GO enrichment analyses.

```{r genes, echo=F}

  all_introns_tidy <- readRDS(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/common_introns.rds"))


  all_genes <- all_introns_tidy  %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  both_misspliced_junc <- all_introns_tidy %>%
    filter(ref_type == "both")
  both_genes <- both_misspliced_junc %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  donor_misspliced_junc <- all_introns_tidy %>%
    filter(ref_type == "donor")
  donor_genes <- donor_misspliced_junc %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  acceptor_misspliced_junc <- all_introns_tidy %>%
    filter(ref_type == "acceptor")
  acceptor_genes <- acceptor_misspliced_junc %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  never_misspliced_junctions <- all_introns_tidy %>%
    filter(ref_type == "never")
  never_genes <- never_misspliced_junctions %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  ## UNIQUE GENES
  both_genes_unique <- setdiff(both_genes, c(never_genes, donor_genes, acceptor_genes) %>% unique()) 
  donor_genes_unique <- setdiff(donor_genes, c(never_genes, both_genes, acceptor_genes) %>% unique())
  acceptor_genes_unique <- setdiff(acceptor_genes, c(never_genes, donor_genes, both_genes) %>% unique())
  never_genes_unique <- setdiff(never_genes, c(both_genes, donor_genes, acceptor_genes) %>% unique())
  
   df <- data.frame("Intron Type" = "both",
                    N_unique_genes = both_genes_unique %>% length())
   df <- rbind(df,
               data.frame("Intron Type" = "donor",
                          N_unique_genes = donor_genes_unique %>% length()))
   df <- rbind(df,
               data.frame("Intron Type" = "acceptor",
                          N_unique_genes = acceptor_genes_unique %>% length()))
   df <- rbind(df,
               data.frame("Intron Type" = "never",
                          N_unique_genes = never_genes_unique %>% length()))
  
  all_genes_unique <- c(both_genes_unique,
                        #donor_genes_unique,
                        #acceptor_genes_unique,
                        never_genes_unique)

  
  df %>% 
    DT::datatable() %>% 
    return()
  
  

```

### 'Both' mis-spliced introns

GO Enrichment analysis (gprofiler2) performed only over the unique subset of genes from introns that were mis-spliced at 'both' splice sites.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_both, message=FALSE, echo=FALSE}

 
  #######################################
  ## GO ENRICHMENT
  #######################################
  
  
  ## 'Both' type genes
  gene_enrichment_both <- gprofiler2::gost(query = both_genes_unique,
                                           custom_bg = all_genes %>% unique(),
                                           organism = "hsapiens",
                                           ordered_query = T,
                                           correction_method = "bonferroni")

  if (!is.null(gene_enrichment_both$result)) {
    
    if (nrow(gene_enrichment_both$result) > 20) {
      
    GO_enrichment <- gene_enrichment_both$result %>% 
     filter(str_detect(source, pattern = "GO")) %>%
     mutate(go_type = str_sub(source, start = 4, end = 5))
     
    GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                 go_id = GO_enrichment$term_id,
                                                                 go_term = GO_enrichment$term_name),
                                         threshold = 0.7)
     
    df_result <- merge(x = gene_enrichment_both$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                       y = GO_enrich_reduc,
                       by.x = "term_id",
                       by.y = "go_id",
                       all.x = T)

    ## JOIN with KEGG results and add pvalues

    rbind(df_result %>%
            as_tibble() %>%
            dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
            group_by(parent_term, source) %>%
            mutate(p_value = p_value %>% max) %>%
            ungroup() %>%
            distinct(parent_term, .keep_all = T),
          df_result %>%
            filter(source == "KEGG")%>%
            dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
            dplyr::rename(parent_term = term_name)) %>%
      arrange(p_value) %>%
      mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
      #filter(source == "KEGG") %>%
      DT::datatable() %>% 
      return()
    
    } else {
      gene_enrichment_both$result%>%
        dplyr::select(term_name, p_value, query_size, intersection_size, source) %>%
        DT::datatable() %>% 
        return()
    }
  } else {
    
    DT::datatable(data = data.frame(result = "No results to show")) %>% 
      return()
  }
  
  

```

### 'Donor' mis-spliced introns

GO Enrichment analysis (gprofiler2) performed only over the unique subset of genes from introns that were mis-spliced ONLY at 'donor' splice sites.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_donor, message=FALSE, echo=F}

 
  #######################################
  ## GO ENRICHMENT
  #######################################
  
  
  ## 'Donor' type genes
  gene_enrichment_donor <- gprofiler2::gost(query = donor_genes_unique,
                                            custom_bg = all_genes %>% unique(),
                                            organism = "hsapiens",
                                            ordered_query = F,
                                            significant = T,
                                            correction_method = "bonferroni")

  if (!is.null(gene_enrichment_donor$result)) {
    
    if (nrow(gene_enrichment_donor$result) > 20) {
      
    GO_enrichment <- gene_enrichment_donor$result %>% 
     filter(str_detect(source, pattern = "GO")) %>%
     mutate(go_type = str_sub(source, start = 4, end = 5))
     
    GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                 go_id = GO_enrichment$term_id,
                                                                 go_term = GO_enrichment$term_name),
                                         threshold = 0.7)
     
    df_result <- merge(x = gene_enrichment_acceptor$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                       y = GO_enrich_reduc,
                       by.x = "term_id",
                       by.y = "go_id",
                       all.x = T)

    ## JOIN with KEGG results and add pvalues

    rbind(df_result %>%
            as_tibble() %>%
            dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
            group_by(parent_term, source) %>%
            mutate(p_value = p_value %>% max) %>%
            ungroup() %>%
            distinct(parent_term, .keep_all = T),
          df_result %>%
            filter(source == "KEGG")%>%
            dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
            dplyr::rename(parent_term = term_name)) %>%
      arrange(p_value) %>%
      mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
      #filter(source == "KEGG") %>%
      DT::datatable() %>% 
      return()
    } else {
      gene_enrichment_donor$result%>%
        dplyr::select(term_name, p_value, query_size, intersection_size, source) %>%
      DT::datatable() %>% 
      return()
    }
    
  } else {
    DT::datatable(data = data.frame(result = "No results to show")) %>% 
      return()
  }
  
  

```

### 'Acceptor' mis-spliced introns

GO Enrichment analysis (gprofiler2) performed only over the unique subset of genes from introns that were mis-spliced ONLY at 'acceptor' splice sites.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_acceptor, message=FALSE, echo=F}

 
  #######################################
  ## GO ENRICHMENT
  #######################################
  
  
  ## 'Acceptor' type genes
  gene_enrichment_acceptor <- gprofiler2::gost(query = acceptor_genes_unique,
                                               custom_bg = all_genes %>% unique(),
                                               organism = "hsapiens",
                                               ordered_query = F,
                                               correction_method = "bonferroni")
  if (!is.null(gene_enrichment_acceptor$result)) {
    
    if(nrow(gene_enrichment_acceptor$result) > 20) {
    GO_enrichment <- gene_enrichment_acceptor$result %>% 
     filter(str_detect(source, pattern = "GO")) %>%
     mutate(go_type = str_sub(source, start = 4, end = 5))
     
    GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                 go_id = GO_enrichment$term_id,
                                                                 go_term = GO_enrichment$term_name),
                                         threshold = 0.7)
     
    df_result <- merge(x = gene_enrichment_acceptor$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                       y = GO_enrich_reduc,
                       by.x = "term_id",
                       by.y = "go_id",
                       all.x = T)

    ## JOIN with KEGG results and add pvalues

    rbind(df_result %>%
            as_tibble() %>%
            dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
            group_by(parent_term, source) %>%
            mutate(p_value = p_value %>% max) %>%
            ungroup() %>%
            distinct(parent_term, .keep_all = T),
          df_result %>%
            filter(source == "KEGG")%>%
            dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
            dplyr::rename(parent_term = term_name)) %>%
      arrange(p_value) %>%
      mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
      #filter(source == "KEGG") %>%
      DT::datatable() %>% 
      return()
    } else {
      gene_enrichment_acceptor$result%>%
        dplyr::select(term_name, p_value, query_size, intersection_size, source) %>%
      DT::datatable() %>% 
      return()
    }
  }  else {
    
    DT::datatable(data = data.frame(result = "No results to show")) %>% 
      return()
  }
  
  

```

### Never mis-spliced introns

GO Enrichment analysis (gprofiler2) performed only over the unique subset of genes from introns that were NEVER mis-spliced.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_never, message=FALSE, echo=F}

  
  #######################################
  ## GO ENRICHMENT
  #######################################

  ## 'Never' type genes
  gene_enrichment_never <- gprofiler2::gost(query = never_genes_unique,
                                            custom_bg = all_genes %>% unique(),
                                            organism = "hsapiens",
                                            ordered_query = F,
                                            correction_method = "bonferroni")
  
  if (!is.null(gene_enrichment_never$result)) {
    
    if(nrow(gene_enrichment_acceptor$result) > 20) {
    GO_enrichment <- gene_enrichment_never$result %>% 
     filter(str_detect(source, pattern = "GO")) %>%
     mutate(go_type = str_sub(source, start = 4, end = 5))
     
    GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                 go_id = GO_enrichment$term_id,
                                                                 go_term = GO_enrichment$term_name),
                                         threshold = 0.7)
     
    df_result <- merge(x = gene_enrichment_never$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                       y = GO_enrich_reduc,
                       by.x = "term_id",
                       by.y = "go_id",
                       all.x = T)

    ## JOIN with KEGG results and add pvalues

    rbind(df_result %>%
            as_tibble() %>%
            dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
            group_by(parent_term, source) %>%
            mutate(p_value = p_value %>% max) %>%
            ungroup() %>%
            distinct(parent_term, .keep_all = T),
          df_result %>%
            filter(source == "KEGG")%>%
            dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
            dplyr::rename(parent_term = term_name)) %>%
      arrange(p_value) %>%
      mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
      #filter(source == "KEGG") %>%
      DT::datatable() %>% 
      return()
    } else {
      gene_enrichment_never$result%>%
        dplyr::select(term_name, p_value, query_size, intersection_size, source) %>%
      DT::datatable() %>% 
      return()
    }
  }  else {
    
   DT::datatable(data.frame(result = "No results to show")) %>%
      return()
  }
  
  
```



## Gene Enrichment - NON Unique genes {.tabset .tabset-fade .tabset-pills} 

### Stats

Due to the small number of genes that were unique to each intron category, we repeated the previous analysis but without removing the overlapping genes.


```{r genes_nonunique, echo=F}

  all_introns_tidy <- readRDS(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/common_introns.rds"))


  all_genes <- all_introns_tidy  %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  both_misspliced_junc <- all_introns_tidy %>%
    filter(ref_type == "both")
  both_genes <- both_misspliced_junc %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  donor_misspliced_junc <- all_introns_tidy %>%
    filter(ref_type == "donor")
  donor_genes <- donor_misspliced_junc %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  acceptor_misspliced_junc <- all_introns_tidy %>%
    filter(ref_type == "acceptor")
  acceptor_genes <- acceptor_misspliced_junc %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  never_misspliced_junctions <- all_introns_tidy %>%
    filter(ref_type == "never")
  never_genes <- never_misspliced_junctions %>% unnest(gene_name) %>% distinct(gene_name) %>% pull()
  
  
  ## UNIQUE GENES
  #both_genes_unique <- setdiff(both_genes, c(never_genes, donor_genes, acceptor_genes) %>% unique()) 
  #donor_genes_unique <- setdiff(donor_genes, c(never_genes, both_genes, acceptor_genes) %>% unique())
  #acceptor_genes_unique <- setdiff(acceptor_genes, c(never_genes, donor_genes, both_genes) %>% unique())
  #never_genes_unique <- setdiff(never_genes, c(both_genes, donor_genes, acceptor_genes) %>% unique())
  
   df <- data.frame("Intron Type" = "both",
                    N_unique_genes = both_genes %>% length())
   df <- rbind(df,
               data.frame("Intron Type" = "donor",
                          N_unique_genes = donor_genes %>% length()))
   df <- rbind(df,
               data.frame("Intron Type" = "acceptor",
                          N_unique_genes = acceptor_genes %>% length()))
   df <- rbind(df,
               data.frame("Intron Type" = "never",
                          N_unique_genes = never_genes %>% length()))

  
  df %>% 
    DT::datatable() %>% 
    return()
  
  

```

### 'Both' mis-spliced introns

GO Enrichment analysis (gprofiler2) performed over genes from introns that were mis-spliced at 'both' splice sites.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_both_nonunique, message=FALSE, echo=FALSE}

 
  #######################################
  ## GO ENRICHMENT
  #######################################
  
  
  ## 'Both' type genes
  gene_enrichment_both <- gprofiler2::gost(query = both_genes,
                                           custom_bg = all_genes %>% unique(),
                                           organism = "hsapiens",
                                           ordered_query = F,
                                           correction_method = "bonferroni")

  if (!is.null(gene_enrichment_both$result)) {
    
    GO_enrichment <- gene_enrichment_both$result %>% 
     filter(str_detect(source, pattern = "GO")) %>%
     mutate(go_type = str_sub(source, start = 4, end = 5))
     
    GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                 go_id = GO_enrichment$term_id,
                                                                 go_term = GO_enrichment$term_name),
                                         threshold = 0.7)
     
    df_result <- merge(x = gene_enrichment_both$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                       y = GO_enrich_reduc,
                       by.x = "term_id",
                       by.y = "go_id",
                       all.x = T)

    ## JOIN with KEGG results and add pvalues

    rbind(df_result %>%
            as_tibble() %>%
            dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
            group_by(parent_term, source) %>%
            mutate(p_value = p_value %>% max) %>%
            ungroup() %>%
            distinct(parent_term, .keep_all = T),
          df_result %>%
            filter(source == "KEGG")%>%
            dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
            dplyr::rename(parent_term = term_name)) %>%
      arrange(p_value) %>%
      mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
      #filter(source == "KEGG") %>%
      DT::datatable() %>% 
      return()
    
  } else {
    
    DT::datatable(data = data.frame(result = "No results to show")) %>% 
      return()
  }
  
  

```

### 'Donor' mis-spliced introns

GO Enrichment analysis (gprofiler2) performed over genes from introns that were mis-spliced at 'donor' splice sites.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_donor_nonunique, message=FALSE, echo=F}

 
  #######################################
  ## GO ENRICHMENT
  #######################################
  
  
  ## 'Donor' type genes
  gene_enrichment_donor <- gprofiler2::gost(query = donor_genes,
                                            custom_bg = all_genes %>% unique(),
                                            organism = "hsapiens",
                                            ordered_query = F,
                                            significant = T,
                                            correction_method = "bonferroni")

  if (!is.null(gene_enrichment_donor$result)) {
    
    if(nrow(gene_enrichment_donor$result) > 20) {
      
    GO_enrichment <- gene_enrichment_donor$result %>% 
     filter(str_detect(source, pattern = "GO")) %>%
     mutate(go_type = str_sub(source, start = 4, end = 5))
     
    GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                 go_id = GO_enrichment$term_id,
                                                                 go_term = GO_enrichment$term_name),
                                         threshold = 0.7)
     
    df_result <- merge(x = gene_enrichment_donor$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                       y = GO_enrich_reduc,
                       by.x = "term_id",
                       by.y = "go_id",
                       all.x = T)

    ## JOIN with KEGG results and add pvalues

    rbind(df_result %>%
            as_tibble() %>%
            dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
            group_by(parent_term, source) %>%
            mutate(p_value = p_value %>% max) %>%
            ungroup() %>%
            distinct(parent_term, .keep_all = T),
          df_result %>%
            filter(source == "KEGG")%>%
            dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
            dplyr::rename(parent_term = term_name)) %>%
      arrange(p_value) %>%
      mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
      #filter(source == "KEGG") %>%
      DT::datatable() %>% 
      return()
    } else {
      gene_enrichment_donor$result%>%
        dplyr::select(term_name, p_value, query_size, intersection_size, source) %>%
      DT::datatable() %>% 
      return()
    }
    
  } else {
    DT::datatable(data = data.frame(result = "No results to show")) %>% 
      return()
  }
  
  

```

### 'Acceptor' mis-spliced introns

GO Enrichment analysis (gprofiler2) performed over genes from introns that were mis-spliced at 'accetor' splice sites.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_acceptor_nonunique, message=FALSE, echo=F}

 
  #######################################
  ## GO ENRICHMENT
  #######################################
  
  
  ## 'Acceptor' type genes
  gene_enrichment_acceptor <- gprofiler2::gost(query = acceptor_genes,
                                               custom_bg = all_genes %>% unique(),
                                               organism = "hsapiens",
                                               ordered_query = F,
                                               correction_method = "bonferroni")
  if (!is.null(gene_enrichment_acceptor$result)) {
    
    if(nrow(gene_enrichment_acceptor$result) > 20) {
      
    GO_enrichment <- gene_enrichment_acceptor$result %>% 
     filter(str_detect(source, pattern = "GO")) %>%
     mutate(go_type = str_sub(source, start = 4, end = 5))
     
    GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                 go_id = GO_enrichment$term_id,
                                                                 go_term = GO_enrichment$term_name),
                                         threshold = 0.7)
     
    df_result <- merge(x = gene_enrichment_acceptor$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                       y = GO_enrich_reduc,
                       by.x = "term_id",
                       by.y = "go_id",
                       all.x = T)

    ## JOIN with KEGG results and add pvalues

    rbind(df_result %>%
            as_tibble() %>%
            dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
            group_by(parent_term, source) %>%
            mutate(p_value = p_value %>% max) %>%
            ungroup() %>%
            distinct(parent_term, .keep_all = T),
          df_result %>%
            filter(source == "KEGG")%>%
            dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
            dplyr::rename(parent_term = term_name)) %>%
      arrange(p_value) %>%
      mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
      #filter(source == "KEGG") %>%
      DT::datatable() %>% 
      return()
    } else {
      gene_enrichment_acceptor$result%>%
        dplyr::select(term_name, p_value, query_size, intersection_size, source) %>%
      DT::datatable() %>% 
      return()
    }
  }  else {
    
    DT::datatable(data = data.frame(result = "No results to show")) %>% 
      return()
  }
  
  

```

### Never mis-spliced introns

GO Enrichment analysis (gprofiler2) performed over genes from introns that were NEVER mis-spliced.

The background of genes used was formed by the ALL 16,003 genes from ALL common introns across GTEx v8 tissues.

The inputed dataset of genes were not ordered and the p-values returned by the GO Enrichment analysis were corrected using the Bonferroni method.

```{r GO_never_nonunique, message=FALSE, echo=F}

  
  #######################################
  ## GO ENRICHMENT
  #######################################

  ## 'Never' type genes
  gene_enrichment_never <- gprofiler2::gost(query = never_genes,
                                            custom_bg = all_genes %>% unique(),
                                            organism = "hsapiens",
                                            ordered_query = F,
                                            correction_method = "bonferroni")
  
  if (!is.null(gene_enrichment_never$result)) {
    
    if(nrow(gene_enrichment_never$result) > 20) {
      
      GO_enrichment <- gene_enrichment_never$result %>% 
       filter(str_detect(source, pattern = "GO")) %>%
       mutate(go_type = str_sub(source, start = 4, end = 5))
       
      GO_enrich_reduc <- rutils::go_reduce(pathway_df = data.frame(go_type = GO_enrichment$go_type,
                                                                   go_id = GO_enrichment$term_id,
                                                                   go_term = GO_enrichment$term_name),
                                           threshold = 0.7)
       
      df_result <- merge(x = gene_enrichment_never$result %>% filter(source %in% c("GO:BP", "GO:MF", "GO:CC", "KEGG")),
                         y = GO_enrich_reduc,
                         by.x = "term_id",
                         by.y = "go_id",
                         all.x = T)
  
      ## JOIN with KEGG results and add pvalues
  
      rbind(df_result %>%
              as_tibble() %>%
              dplyr::select(parent_term, p_value, query_size, intersection_size, source) %>%
              group_by(parent_term, source) %>%
              mutate(p_value = p_value %>% max) %>%
              ungroup() %>%
              distinct(parent_term, .keep_all = T),
            df_result %>%
              filter(source == "KEGG")%>%
              dplyr::select(p_value, query_size, intersection_size, source, term_name) %>%
              dplyr::rename(parent_term = term_name)) %>%
        arrange(p_value) %>%
        mutate(p_value = p_value %>% formatC(format = "e", digits = 2)) %>%
        #filter(source == "KEGG") %>%
        DT::datatable() %>% 
        return()
      
    } else {
      
      gene_enrichment_never$result%>%
        dplyr::select(term_name, p_value, query_size, intersection_size, source) %>%
      DT::datatable() %>% 
      return()
      
    }
  }  else {
    
   DT::datatable(data.frame(result = "No results to show")) %>%
      return()
  }
  
  
```


