---
title: "RBP analyses"
author: "Sonia Garcia-Ruiz"
date: "13/05/2022"
output: html_document
---

```{r setup, include=FALSE}



knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(DESeq2)
library(SummarizedExperiment)
library(biomaRt)

```

# Intro

RNA binding proteins (RBPs) have emerged as central players in regulating gene expression, controlling when, where, and at what rate RNAs are processed, trafficked, translated, and degraded within the cell. 

For regulation of alternative splicing, where RBP binding can cause either inclusion or exclusion of alternative exons, it is common to identify RBP-responsive events by knocking down or over-expressing the RBP and performing RNA-seq. 

*In 2020 Van Nostrand et al. published a study reporting the results of studying the functions of 356 human RBPs using integrative approaches consisting of 5 assays (eCLIP, RNA Bind-N-Seq, Immunofluorescence, Knockdown RNA-seq, RBP ChIP-Seq) that focused on different aspects of RBP activity. These hundred and fifty-six RBPs had also been profiled by at least one ENCODE experiment.* [A Large-Scale Binding and Functional Map of Human RNA Binding Proteins](https://www.nature.com/articles/s41586-020-2077-3).

Functionally, these 356 RBPs are most commonly known to play roles in the regulation of RNA splicing (98 RBPs, 28%), RNA stability and decay (71, 20%), and translation (70, 20%), with 162 RBPs (46%) having more than one function reported in the literature.

**In this study, we have made use of these 356 RBPs to study their changes in expression levels across age groups and body parts.**

# Pipeline

The image below contains (1) the approach followed to calculate the expression levels (i.e. TPM values) corresponding to each RPB per sample, and (2) and the pipeline run in the two main analyses presented in this document.

![Image 1. Complete pipeline followed to performed the analyses shown in this document.](rbp_expression_pipeline_analysis.png)

# Heatmap {.tabset .tabset-fade .tabset-pills}

TPM values graphically represented in this section have been corrected by the following covariates: age, center, gebtch, gebtchd, nabtc, nabtchd, nabtcht, hhrdy, sex, rin, by following [Fairbrother-Browne, A. et.al](https://www.nature.com/articles/s42003-021-02792-w) approach.


## Blood
![Image 2. RBP mean expression levels across all samples from each age group in BLOOD tissue.](blood_heatmap.png)

## Brain

![Image 3. RBP mean expression levels across all samples from each age group in BRAIN tissue.](brain_heatmap.png)


## Muscle

![Image 4. RBP mean expression levels across all samples from each age group in MUSCLE tissue.](muscle_heatmap.png)

# Linear Model comparisons {.tabset .tabset-fade .tabset-pills}



## Results

The table below contains all RBPs whose uncorrected TPM value was significantly affected by age in each one of the three body parts studied.

**TPM ~ β0 + ageβ1 + rinβ2 + sexβ3 + centerβ4 + tissueβ5 + [btch_VariablesβN]**

Each value shown per RBP and tissue corresponds to the effect that age had over the uncorrected TPM value of that RBP in that particular tissue (i.e. it is the *beta* value for age returned by the linear model).

```{r lm_comparison, echo=FALSE}
  projects_id <- c("BRAIN", "MUSCLE", "BLOOD")
  
  df_tissues <- map_df(projects_id, function(project_id) {
    
    # project_id <- projects_id[1]
    
    df_analysis_lm_uncorrected <- read.csv(file = paste0("/home/sruiz/PROJECTS/splicing-project/splicing-recount3-projects/", project_id,
                                                         "/results/pipeline3/rbp/tpm_lm.csv")) %>%
      mutate(tissue = project_id)
    
    return(df_analysis_lm_uncorrected)
    
  })
  
  df_tissues %>%
    dplyr::select(Estimate, hgnc_symbol, tissue) %>%
    spread(key = tissue, value = Estimate) %>%
    dplyr::rename(RBP = hgnc_symbol) %>%
    DT::datatable()
  
```

## Stats

The table below contains all RBPs whose uncorrected TPM value was significantly affected by age in each one of the three body parts studied.
The column "n_tissues" shows whether the RBP overlaps across all tissues ( n_tissues = 3 ) or it is only particular to one of the three tissues.

```{r lm_comparison_stats, echo=FALSE}

df_tissues %>%
  group_by(hgnc_symbol) %>%
  dplyr::mutate(n_tissues = n()) %>%
  distinct(hgnc_symbol, .keep_all = T) %>% 
  dplyr::select(hgnc_symbol, n_tissues) %>%
  dplyr::arrange(desc(n_tissues),hgnc_symbol) %>%
  DT::datatable()

  
```

## Blood

The table below is a subset of the previous table. It contains only the RBPs that are unique to BLOOD tissue (i.e. that have shown to be significantly affected by age only in BLOOD):

```{r lm_comparison_stats_blood, echo=FALSE}


df_tissues %>%
  group_by(hgnc_symbol) %>%
  dplyr::mutate(n_tissues = n()) %>%
  filter(n_tissues == 1,
         tissue == "BLOOD") %>%
  dplyr::select(hgnc_symbol, age_effect = Estimate) %>%
  arrange(hgnc_symbol) %>%
  DT::datatable()


  
```

## Brain

The table below is a subset of the previous table. It contains only the RBPs that are unique to BRAIN tissue (i.e. that have shown to be significantly affected by age only in BRAIN):

```{r lm_comparison_stats_brain, echo=FALSE}


RBPs_brain <- df_tissues %>%
  group_by(hgnc_symbol) %>%
  dplyr::mutate(n_tissues = n()) %>%
  filter(n_tissues == 1,
         tissue == "BRAIN") %>%
  dplyr::select(hgnc_symbol, age_effect = Estimate) %>%
  arrange(hgnc_symbol) 

RBPs_brain %>%
  DT::datatable()


  
```



## Muscle 

The table below is a subset of the previous table. It contains only the RBPs that are unique to MUSCLE tissue (i.e. that have shown to be significantly affected by age only in MUSCLE):

```{r lm_comparison_stats_muscle, echo=FALSE}


df_tissues %>%
  group_by(hgnc_symbol) %>%
  dplyr::mutate(n_tissues = n()) %>%
  filter(n_tissues == 1,
         tissue == "MUSCLE") %>%
  dplyr::select(hgnc_symbol, age_effect = Estimate) %>%
  arrange(hgnc_symbol) %>%
  DT::datatable()


  
```